ods-ci 0502__ide_elyra.robot: suite setup fails, "Verify Pipeline Server Deployments" — Pending != Running (kube-rbac-proxy ImagePullBackOff, missing image override for DSPA v2.18.0)

## Where

- Run: https://github.com/jiridanek/rhoai-in-kind/actions/runs/31151019542
- Job: `rhoai-in-kind (v1.36.0, tests/Tests/0500__ide/0502__ide_elyra.robot)` — https://github.com/jiridanek/rhoai-in-kind/actions/runs/31151019542/job/92780481292?pr=36
- Failing step: `Run ods-ci tests` (exit code 2)
- PR: #36 (`jd_225` — rhoai-2.25 baseline pin)

## What happened (from the raw job log)

Both tests in the suite fail at **suite setup**, not inside the actual Elyra pipeline logic:

```
Verify Pipelines Integration With Elyra When Using Standard Data S... | FAIL |
Parent suite setup failed:
Keyword 'Verify Pipeline Server Deployments' failed after retrying 15 times. The last error was: Pending != Running
...
Ide Elyra :: Test Suite for Elyra pipelines in workbenches            | FAIL |
Suite setup failed:
Keyword 'Verify Pipeline Server Deployments' failed after retrying 15 times. The last error was: Pending != Running

2 tests, 0 passed, 2 failed
```

`output.xml` confirms the same `Verify Pipeline Server Deployments` keyword (library `DataSciencePipelinesBackend`) retries and fails with `Pending != Running` every ~10s from 05:40:57 through the final failure — this is a k8s-API-level check (not a Selenium/UI wait), so it's polling actual `Deployment`/pod status directly.

## Root cause (confirmed via cluster-logs artifact)

`cluster-logs-0502__ide_elyra/namespaces/elyra-test-8459/core/podss.yaml` and `eventss.yaml` show two of the DSPA pods permanently stuck `phase: Pending`:

- `ds-pipeline-dspa-67cc9d4c8b-x6tmf` (the `ds-pipeline-api-server` deployment)
- `ds-pipeline-metadata-envoy-dspa-fffd68dbc-nsghn`

Both have a `kube-rbac-proxy` sidecar container in `ImagePullBackOff`. The event log gives the exact failure:

```
message: 'Failed to pull image "registry.redhat.io/openshift4/ose-kube-rbac-proxy-rhel9@sha256:eac5ade5fc49b724ef33c05c81bf960f7bec3f52faa2e44db249d05bcc45c280":
  failed to pull and unpack image ...:
  failed to resolve reference ...:
  failed to authorize: failed to fetch anonymous token: unexpected status from GET
  request to https://registry.redhat.io/auth/realms/rhcc/protocol/redhat-docker-v2/auth?scope=repository%3Aopenshift4%2Fose-kube-rbac-proxy-rhel9%3Apull&service=docker-registry:
  401 Unauthorized'
reason: Failed
```

This kind cluster has no `registry.redhat.io` pull credentials, so any pod that references a `registry.redhat.io` image directly gets a permanent `401`/`ImagePullBackOff` and never becomes `Ready`. Since two of the DSPA's pods (api-server + metadata-envoy) never come up, `Verify Pipeline Server Deployments` can never see `Running` and the whole suite setup times out. A screenshot taken during the run (`selenium-screenshot-7.png`) shows the dashboard correspondingly stuck on the project's Pipelines tab at **"Initializing Pipeline Server... This may take a while."**

**This is not a new/unknown problem class for this repo** — `components/03-kf-pipelines.yaml` already has an established workaround for exactly this class of issue: it patches the `data-science-pipelines-operator-controller-manager` Deployment's env vars (`IMAGES_MLMDENVOY`, `IMAGES_OAUTHPROXY`, `IMAGES_MARIADB`) to replace `registry.redhat.io` images with public `quay.io`/`quay.io/fedora` mirrors, with a comment referencing an internal Slack thread about registry access. That patch was added in commit `2ec63d9` ("pin odh-dashboard and DSPA to specific versions...", part of this same PR #36 pinning effort), which also pinned `targetRevision: v2.18.0` for the DSP operator.

The gap: I fetched the operator's `config/base/params.env` at the pinned tag `v2.18.0` directly from GitHub and it defines:

```
kube-rbac-proxy=registry.redhat.io/openshift4/ose-kube-rbac-proxy-rhel9@sha256:eac5ade5fc49b724ef33c05c81bf960f7bec3f52faa2e44db249d05bcc45c280
```

— the exact same image digest seen failing in the events above. This env var was not included in the `03-kf-pipelines.yaml` override patch, most likely because it doesn't follow the `IMAGES_*` naming convention used by all the other overridden vars (`IMAGES_MLMDENVOY`, `IMAGES_OAUTHPROXY`, `IMAGES_MARIADB`), so it's easy to miss when grepping/eyeballing the override list.

## Suggested fix (not applied here, per instructions to only analyze)

Add `kube-rbac-proxy` to the `env` patch block in `components/03-kf-pipelines.yaml` (around lines 61-69), pointing at a public mirror of `ose-kube-rbac-proxy-rhel9` (there may already be one used elsewhere in this repo for the same/similar kube-rbac-proxy image used by other operators — worth checking before picking a new mirror).

## What I could not confirm

- Whether *only* these two pods (api-server, metadata-envoy) reference `kube-rbac-proxy`, or whether other DSPA-managed pods/containers in this deployment also pull it under a different digest/tag that would need the same override — I only inspected the pods that were actually `Pending` in this run's artifact.
- Whether this exact combination (two specific sidecars stuck) is deterministic every run, since retries/scheduling order could in principle leave a different subset of pods `Pending` on a different run — but the underlying cause (missing `kube-rbac-proxy` override) would apply regardless of which pod hits it first.

cc @coderabbitai — flagging in case you spot another `IMAGES_*`-style env var in the v2.18.0 operator that also needs a mirror, or think the mirror-selection approach here is wrong.

Root-caused the specific failure of `Verify Pipelines Integration With Elyra When Using Standard Data Science Image` (the Sanity test in `0502__ide_elyra.robot`) from run [95886583964](https://github.com/jiridanek/rhoai-in-kind/actions/runs/32191489985/job/95886583964) — it failed at 22:26:28 with `Element '...text()='Succeeded'' did not appear in 5 minutes`, and the pipeline-run screenshot showed the KFP UI status as **`Failed`** with the very first node, `load-weather-data`, marked with a red error icon (distinct from, and earlier than, the disk-I/O-saturation issue in #67 — that job's first `dmesg` hung-task warning doesn't appear until 22:41, ~15 minutes after this test already failed).

## Root cause

`load-weather-data` runs [`load_data.ipynb`](https://github.com/redhat-rhods-qe/ods-ci-notebooks-main/blob/main/notebooks/500__jupyterhub/pipelines/v2/elyra/run-pipelines-on-data-science-pipelines/load_data.ipynb), whose only real cell does:

```python
r = requests.get(data_file)  # data_file = a github.com/.../raw/... tarball URL
if r.status_code != 200:
    raise RuntimeError(...)
```

**No `timeout=` argument.** The follow-on notebook, `Part 1 - Data Cleaning.ipynb`, similarly runs `!pip3 install PyGithub pandas` as its first cell with no timeout on that implicit network call either.

This is a real, easily reproducible fragility bug: if the pod's network path to GitHub/PyPI has *any* transient hiccup, a call with no timeout either hangs forever (if the hiccup is "connection never completes") or raises a fast, opaque `RuntimeError`/pipeline `Failed` (if the hiccup is "got a bad/no HTTP response") — instead of a clear, retryable, fast error either way.

## Live local reproduction

While investigating, found a background pipeline-run pod in `elyra-repro3` that had been stuck 8+ hours (leftover from an earlier manual repro session), still `Running`. Its `main` container log was frozen mid-execution:

```
[I 23:01:12.727] Executing notebook with kernel: python3
Executing:   2%|▏  | 1/47 [00:00<00:45,  1.02cell/s][D 23:01:12.763] Executing cell:
!pip3 install PyGithub pandas > /dev/null 2>&1
```

...with zero log growth over a subsequent 15s poll. `kubectl exec` into it confirmed the mechanism precisely:

```
$ getent hosts pypi.org
151.101.192.223 pypi.org   # DNS resolves fine
$ curl -m 8 https://pypi.org
# exit code 124 — timed out, no connection, no RST, just silence
```

DNS resolves but the TCP connection to the resolved IP is silently black-holed. (Side note: the pod's own `elyra/node-name` label said `load-weather-data`, but that's a mislabeling — matching the tar.gz artifact UUID against the actual `hello-generic-world.pipeline` definition proved it was really executing `Part 1 - Data Cleaning.ipynb`. This mislabeling reproduces on every run.)

**Then reproduced it live, in a completely fresh run**, after cleaning up the stale pod: submitted the pipeline via the real JupyterLab "Run Pipeline" dialog against a freshly-started workbench. `load-weather-data` itself succeeded in under a minute this time. But the very next node — again mislabeled `load-weather-data`, again actually `Part 1 - Data Cleaning.ipynb` — hung on the identical `!pip3 install PyGithub pandas` line, confirmed stuck via 7 consecutive log polls over 4.5 minutes with zero output growth.

## Why this matters beyond "flaky local network"

Digging into *why* pod egress is unreliable turned up something broader than "no internet from pods": CoreDNS's own logs show it intermittently timing out **reaching the in-cluster API server**, not just external DNS:

```
[ERROR] plugin/kubernetes: ...: dial tcp 10.96.0.1:443: i/o timeout
[ERROR] plugin/errors: 2 pypi.org.redhat.com. A: read udp 10.244.0.11:38693->10.89.0.1:53: i/o timeout
```

i.e. this is intermittent packet loss on the kind pod overlay network severe enough to affect cluster-internal ClusterIP traffic, not merely a "blocked egress" firewall rule — confirmed the podman-machine host and a default-podman-network container reach `pypi.org` fine every time; only traffic through kind's pod network (kindnet CNI, on this macOS/podman-machine host) is affected.

This local-network flakiness is almost certainly host/environment-specific (macOS + podman machine + kindnet) and won't manifest identically on the GitHub Actions runner. But the **portable, CI-relevant bug** is the notebooks' lack of any timeout — that's a real fragility regardless of environment, and a plausible (if unconfirmed — the run's cluster-logs artifact came back empty, see #67) explanation for the original CI failure too.

## Suggested fix

- Upstream (`redhat-rhods-qe/ods-ci-notebooks-main`): add `timeout=` to the `requests.get()` call in `load_data.ipynb`, and consider a timeout/retry wrapper around the `pip3 install` cells used across these notebooks, so a network hiccup produces a fast, clear, retryable failure instead of an indefinite hang or an opaque `RuntimeError`.
- Separately worth investigating: the `elyra/node-name` pod label consistently pointing at the wrong pipeline node (reproduced on two independent runs) makes debugging these pods harder than necessary — worth a look upstream in Elyra/KFP-launcher's pod-labeling code, though that's secondary to the timeout issue.

(Local cleanup done: deleted the 8h-stale workflow/pod, removed a stuck `notebook-oauth-client-finalizer.opendatahub.io` finalizer that had an unrelated namespace wedged in `Terminating` for 11h, and tore down the fresh repro run afterward too.)

cc @coderabbitai — any thoughts on whether it's worth adding a client-side timeout/retry wrapper in this repo's own test setup (e.g. pre-warming DNS/egress before the Elyra suite runs), versus this being purely something to fix upstream in the notebooks themselves?

--- a/components/03-kf-pipelines.yaml
+++ b/components/03-kf-pipelines.yaml
@@ -30,7 +30,9 @@ spec:
     path: config/overlays/rhoai
     # can't use stable-2.x because there images aren't pinned, use :latest
     #  https://github.com/opendatahub-io/data-science-pipelines-operator/blob/stable-2.x/config/base/params.env
-    targetRevision: v2.18.0
+    # v2.15.1 is what rhoai-2.25 actually ships, per
+    #  https://github.com/opendatahub-io/opendatahub-community/issues/183#issuecomment-3351838012
+    targetRevision: v2.15.1
     kustomize:
       patches:
         - target:

--- a/components/09-kf-notebooks/kustomization.yaml
+++ b/components/09-kf-notebooks/kustomization.yaml
@@ -21,10 +21,18 @@ resources:
   #- git@github.com:opendatahub-io/kubeflow.git/components/notebook-controller/config/overlays/openshift/
   #- git@github.com:opendatahub-io/kubeflow.git/components/odh-notebook-controller/config/base/
 
+  # v1.10.0-6 (the community tag matching opendatahub-community#183's "ODH 2.35.0 ~ rhoai-2.25"
+  # tracker) already contains RHOAIENG-32633 (renamed the Elyra runtime config's display_name
+  # from "Data Science Pipeline" to "AI Pipeline"), which Jira targets at rhoai-3.0, not 2.25 -
+  # it landed on main days before the community tag was cut, but isn't in the downstream 2.x
+  # scope. Verified directly against the real downstream red-hat-data-services/kubeflow
+  # `rhoai-2.25` branch: both controllers are pinned there to v1.10.0-5, and that build still
+  # says "Data Science Pipeline" (matching ods-ci's release-2.25 suite).
   # https://github.com/opendatahub-io/kubeflow/releases/tag/v1.10.0-6
   # https://github.com/opendatahub-io/opendatahub-community/issues/183#issuecomment-3339632379
-  - https://github.com/opendatahub-io/kubeflow//components/notebook-controller/config/overlays/openshift?ref=v1.10.0-6&submodules=false
-  - https://github.com/opendatahub-io/kubeflow//components/odh-notebook-controller/config/base?ref=v1.10.0-6&submodules=false
+  # https://issues.redhat.com/browse/RHOAIENG-32633
+  - https://github.com/opendatahub-io/kubeflow//components/notebook-controller/config/overlays/openshift?ref=v1.10.0-5&submodules=false
+  - https://github.com/opendatahub-io/kubeflow//components/odh-notebook-controller/config/base?ref=v1.10.0-5&submodules=false
   - culler-config.yaml
 
 patches:

